src/python/zquantum/core/testing/generate_cases_for_backend_tests.py

```diff
@@ -0,0 +1,333 @@
+    for initial_matrix, initial_matrix_names in two_qubit_initial_states:
+        outputs = []
+        operator_names = []
+        for operator, operator_name in two_qubit_operators:
+            operator_names.append('"{}"'.format(operator_name))
+            circuit_ket = mul(matrix, initial_matrix) * sympy.Matrix(
+                [[1], [0], [0], [0]]
+            )
+            circuit_bra = sympy.conjugate(sympy.Transpose(circuit_ket))
+            expectation_value = mul(circuit_bra, mul(operator, circuit_ket))[0]
+
+            outputs.append(sympy.simplify(expectation_value))
+        operator_names_string = "[" + ", ".join(operator_names) + "], "
+        gate_names_string = '[["{}", "{}"], "{}", '.format(
+            initial_matrix_names[0], initial_matrix_names[1], matrix_name
+        )
+        exp_vals_string = "["
+        for output in outputs:
+            exp_vals_string += (
+                "{},".format(output)
+                .replace("sqrt", "np.sqrt")
+                .replace("pi", "np.pi")
+                .replace("1.0*I", "1.0j")
+                .replace("*I", "*1.0j")
+            )
+        exp_vals_string += "]"
+
+        print(gate_names_string + operator_names_string + exp_vals_string + "],")
+
+
+def generate_cases_2_qubits_exp_vals_with_angles(matrix, matrix_name, angles):
+    for initial_matrix, initial_matrix_names in two_qubit_initial_states:
+        for angle in angles:
+            outputs = []
+            operator_names = []
+            for operator, operator_name in two_qubit_operators:
+                operator_names.append('"{}"'.format(operator_name))
+                circuit_ket = mul(matrix, initial_matrix) * sympy.Matrix(
+                    [[1], [0], [0], [0]]
+                )
+                circuit_ket = circuit_ket.subs("theta", angle)
+                circuit_bra = sympy.conjugate(sympy.Transpose(circuit_ket))
+                expectation_value = mul(circuit_bra, mul(operator, circuit_ket))[0]
+
+                outputs.append(sympy.simplify(expectation_value))
+            operator_names_string = "[" + ", ".join(operator_names) + "], "
+            gate_names_string = '[["{}", "{}"], "{}", '.format(
+                initial_matrix_names[0], initial_matrix_names[1], matrix_name
+            )
+            angle_string = "[{}], ".format(angle).replace("pi", "np.pi")
+            exp_vals_string = "["
+            for output in outputs:
+                exp_vals_string += (
+                    "{},".format(output)
+                    .replace("sqrt", "np.sqrt")
+                    .replace("pi", "np.pi")
+                    .replace("1.0*I", "1.0j")
+                    .replace("*I", "*1.0j")
+                )
+            exp_vals_string += "]"
+
+            print(
+                gate_names_string
+                + angle_string
+                + operator_names_string
+                + exp_vals_string
+                + "],"
+            )
```

Comment:
It's hard to really know what's going on here. Maybe add some comments above each function saying what kinds of cases it is generating?

Reply:
I actually think that the names of the functions are pretty clear, do you have a suggestion what should I add to make it clearer?

Reply:
I was confused because I didn't understand why you were building strings and then printing them out without returning anything. When reading this, I actually initially thought it was creating some global variables or something to be used during testing. Now I realize these are those functions you used to generate the code for the different testing states. Maybe add a comment at the top of the file describing that this is what these functions are for and maybe when/why/how someone would alter them and/or use them?

Reply:
Addressed in [46e52f](https://github.com/zapatacomputing/z-quantum-core/pull/145/commits/46e52f81e7ca383dc39a0b581a221a5a3c2693c3).